block-producer/src/store/mod.rs

```diff
@@ -0,0 +1,26 @@
+use std::collections::BTreeMap;
+
+use async_trait::async_trait;
+use miden_objects::Digest;
+
+use crate::SharedProvenTx;
+
+#[derive(Debug)]
+pub enum TxInputsError {}
+
+/// Information needed from the store to verify a transaction
+pub struct TxInputs {
+    /// The account hash in the store corresponding to tx's account ID
+    pub account_hash: Option<Digest>,
+
+    /// Maps each consumed notes' nullifier to whether the note is already consumed
+    pub nullifiers: BTreeMap<Digest, bool>,
+}
+
+#[async_trait]
+pub trait GetTxInputs: Send + Sync + 'static {
+    async fn get_tx_inputs(
+        &self,
+        proven_tx: SharedProvenTx,
+    ) -> Result<TxInputs, TxInputsError>;
+}
```

Comment:
Done. I had to keep the `ApplyBlock` trait separate (and [make a supertrait](https://github.com/0xPolygonMiden/miden-node/blob/540516a3136b21f8a5f406ce55d97cbe2f528e91/block-producer/src/store/mod.rs#L32) of `Store`), since `StateView` only needs to implement that method (and not the whole `Store` trait)

Reply:
Overall, we have 3 endpoints which we'll need to implement on the store: `apply_block`, `get_block_inputs` and `get_tx_inputs`. `StateView` will need `apply_block` and `get_tx_inputs`, while `BlockBuilder` will need `get_block_inputs`.

We could have 2 or 3 traits to define these - but I think `get_tx_inputs` and `apply_block` would need to be on the same trait.

Reply:
Looking more at the code, I think we may be missing two things here. There needs to be `apply_block()` on the `StateView` and `apply_block()` on the `Store`. Currently, these methods are identical - but they may not be in the future. So, I would probably separate them. One structure we could go with is as follows:

```Rust
pub trait TransactionVerifier {
    async fn verify_tx(
        &self,
        tx: SharedProvenTx,
    ) -> Result<(), VerifyTxError>;
}

// though, maybe we don't need to make this a super trait
pub trait StateView: TransactionVerifier {
    async fn apply_block(
        &self,
        block: Arc<Block>,
    ) -> Result<(), ApplyBlockError>;
}

pub trait Store {
    async fn get_tx_inputs(
        &self,
        proven_tx: SharedProvenTx,
    ) -> Result<TxInputs, TxInputsError>;

    // This could also be moved into a separate trait
    async fn get_block_inputs(
        &self,
        block_sketch: BlockSketch,
    ) -> Result<BlockInputs, BlockInputsError>;

    async fn apply_block(
        &self,
        block: Arc<Block>,
    ) -> Result<(), ApplyBlockError>;
}
```

Then we could implement `TransactionVerifier` and `StateView` on `DefaultStateView` and `Store` on `DefaultStore`.